Samples: Bluetooth: Add ST BLE Demo sample

mtuxpe opened the pull request:
This sample demonstrates BLE peripheral for ST BLE Sensor.
You can test button notification and LED service using
ST BLE Sensor Android application.

Signed-off-by: Marcio Montenegro <mtuxpe@gmail.com>

mtuxpe committed:
Samples: Bluetooth: Add ST BLE Demo sample

This sample demonstrates BLE peripheral for ST BLE Sensor.
You can test button notification and LED service using
ST BLE Sensor Android application

Signed-off-by: Marcio Montenegro <mtuxpe@gmail.com>

dbkinder commented on samples/bluetooth/peripheral_nucleo_wb55rg/README.rst:
```diff
@@ -0,0 +1,31 @@
+Overview
+********
+
+Application demonstrating BLE peripheral on `ST Nucleo WB55RG`_
```
Let's make this a sentence, something like:
```
This application demonstrates BLE peripherals on an `ST Nucleo WB55RG`_
by exposing a battery service and vendor-specific GATT services.
```

mtuxpe replied:
done

dbkinder commented on samples/bluetooth/peripheral_nucleo_wb55rg/README.rst:
```diff
@@ -0,0 +1,31 @@
+Requirements
+************
+
+* ST BLE Sensor Android app `ST BLE Sensor app`_
```
Better would be:
```
* `ST BLE Sensor Android app <ST BLE Sensor app>`_
```

mtuxpe replied:
fixed

dbkinder commented on samples/bluetooth/peripheral_nucleo_wb55rg/README.rst:
```diff
@@ -0,0 +1,31 @@
+
+* ST BLE Sensor Android app `ST BLE Sensor app`_
+* A board with STM32WB55 like ST Nucleo WB55RG
+* Zephyr kernel environment configured for ST Nucleo WB55RG `ST Nucleo WB55RG port`_
```
Since this is a zephyr doc, we can reference it with the label at the top of its .rst file:
```
* :ref:`Zephyr kernel environment configured for ST Nucleo WB55RG <nucleo_wb55rg_board>`
```

mtuxpe replied:
fixed

dbkinder commented on samples/bluetooth/peripheral_nucleo_wb55rg/README.rst:
```diff
@@ -0,0 +1,30 @@
+This sample can be found under :zephyr_file:`samples/bluetooth/peripheral_nucleo_wb55rg` in the
+Zephyr tree.
+
+west build -b nucleo_wb55rg samples/bluetooth/peripheral_nucleo_wb55rg
```
Sorry, didn't comment on this earlier.  These build/flash commands should be written as:
```
.. zephyr-app-commands::
   :zephyr-app: samples/bluetooth/peripheral_nucleo_wb55rg
   :board: nucleo_wb55rg
   :goals: build flash
   :compact:

```

mtuxpe replied:
done

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/README.rst:
```diff
@@ -0,0 +1,30 @@
+.. _bluetooth-peripheral_nucleo_wb55rg:
```
Let's rename sample to st_ble_sensor, as this is actually more linked to the mobile application than the board you've used.

mtuxpe replied:
done

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/README.rst:
```diff
@@ -0,0 +1,30 @@
+.. _bluetooth-peripheral_nucleo_wb55rg:
+
+Bluetooth: STM32WB55 BLE peripheral demo
```
Remove references to the board

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/README.rst:
```diff
@@ -0,0 +1,30 @@
+
+This application demonstrates BLE peripherals on an `ST Nucleo WB55RG`_
+by exposing a battery service and vendor-specific GATT services.
+
```
Application proposes a variety of use cases, mostly towards sensors.
So please expose here what are the services currently implemented.

mtuxpe replied:
done

mtuxpe replied:
This application demonstrates BLE peripherals on an `ST Nucleo WB55RG`_
by exposing vendor-specific GATT services.
Currently only button notification and LED service are implemented.

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/prj.conf:
```diff
@@ -0,0 +1,13 @@
+CONFIG_BT_GATT_BAS=y
+CONFIG_BT_GATT_CLIENT=y
+CONFIG_BT_DEBUG=y
+CONFIG_GPIO=y
```
GPIO should not be needed. This is part of the few drivers part of default board configuration

mtuxpe replied:
done

mtuxpe replied:
ok

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/prj.conf:
```diff
@@ -0,0 +1,13 @@
+CONFIG_BT_GATT_CLIENT=y
+CONFIG_BT_DEBUG=y
+CONFIG_GPIO=y
+#CONFIG_BT_DEBUG_HCI_DRIVER=y
```
Remove this line

mtuxpe replied:
fixed

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+sample:
+  description: TBD
```
Will need to be updated

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/src/main.c:
```diff
@@ -0,0 +1,298 @@
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/services/bas.h>
+#include <bluetooth/controller.h>
```
Is `controller.h` actually requested ?

mtuxpe replied:
fixed

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/src/main.c:
```diff
@@ -0,0 +1,298 @@
+
+LOG_MODULE_REGISTER(main);
+
+/* Custom Service Variables */
```
Can you elaborate ?

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/src/main.c:
```diff
@@ -0,0 +1,298 @@
+static struct bt_conn *p_conn;
+
+static u16_t but_val;
+static volatile u8_t notifyEnable;
```
Please avoid CamelCase: `notify_enable`.

mtuxpe replied:
done

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/src/main.c:
```diff
@@ -0,0 +1,298 @@
+
+void led_init(void)
+{
+
```
Remove this line

mtuxpe replied:
done

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/src/main.c:
```diff
@@ -0,0 +1,298 @@
+		LOG_ERR("LED dev error\n");
+		return;
+	}
+	/* Set LED pin as output */
```
Break a line here

mtuxpe replied:
fixed

erwango commented on samples/bluetooth/peripheral_nucleo_wb55rg/src/main.c:
```diff
@@ -0,0 +1,298 @@
+	.disconnected = disconnected,
+};
+
+static void bas_notify(void)
```
Haven't been able to get this service working (tested on disco_l475_iot1)

mtuxpe replied:
On Android ST BLE Sensor app the battery service is not working.
Tested on 4.4.4 version .Always showing "Battery not available"
Even using original STM32CubeIDE p2psrv example.
Android app issue? I will check on ST forum

erwango replied:
> On Android ST BLE Sensor app the battery service is not working.

Well, then we should remove this part. We can keep the placeholder thought for future use if that makes sense.

mtuxpe replied:
battery service removed

erwango reviewed (changes requested):
Thanks for this proposal!
First batch of comments. Main point is that application supports a wide range of services so we'll need to have the application allowing to support these. This is not blocker for upstream since services could be added step by step. Though, we need to make this easy to add services, so we should be able to identify easily what to do to add more services. In that sense some comments would help, aslo it would be nice to have one main file for the application and one file for each service.

mtuxpe commented on samples/bluetooth/peripheral_nucleo_wb55rg/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(eddystone)
```
done

mtuxpe replied:
fixed

dbkinder commented on samples/bluetooth/st_ble_sensor/README.rst:
```diff
@@ -0,0 +1,37 @@
+* `ST BLE Sensor Android app <ST BLE Sensor app>`_
+* A board with STM32WB55 like ST Nucleo WB55RG
+* This application requests having the stm32wb5x_BLE_Stack_fw.bin binary flashed on the Wireless Coprocessor.
+You need to use <STM32CubeProgrammer>`_ to load the appropriate binary.
```
Fix the reference link, and I'm assuming this is a new bullet point, so change to:
```
* Use `STM32CubeProgrammer`_ to load the appropriate binary.
```

mtuxpe replied:
I moved 1 bullet to Overview and added STM32CubeProgrammer bullet to Requirements

jhedberg reviewed (changes requested):
Wouldn't `samples/boards/nucleo_wb55rg/ble_sensor` be a more appropriate location for this?

mtuxpe commented:
> Wouldn't `samples/boards/nucleo_wb55rg/ble_sensor` be a more appropriate location for this?
To be honest I tested this app only on ST Nucleo WB55RG hardware
Probably should work using another platform. I will test in Particle Xenon later.

mtuxpe commented:
> > Wouldn't `samples/boards/nucleo_wb55rg/ble_sensor` be a more appropriate location for this?
> > To be honest I tested this app only on ST Nucleo WB55RG hardware
> > Probably should work using another platform. I will test in Particle Xenon later.

Hi , I confirm that this app also works for non ST hardware.
The ST BLE Android app checks BLE advertising field and ST custom services.

I can't find documentation for the ST custom service UUID ,led service and button notification.
( I used reverse engineering )

The BLE characteristic sensors are documented 
https://www.st.com/resource/en/user_manual/dm00550659.pdf

jhedberg commented:
If you want this to be more generic than just for a single board, please update the documentation, commit message and sample name to reflect that. Currently when looking at those this comes across as being completely related to the WB55RG, which is why I suggested the move to the board-specific samples.

mtuxpe commented:
> If you want this to be more generic than just for a single board, please update the documentation, commit message and sample name to reflect that. Currently when looking at those this comes across as being completely related to the WB55RG, which is why I suggested the move to the board-specific samples.

Yes.I need to review documentation but I changed the commit message:
Samples: Bluetooth: Add ST BLE Demo sample
https://github.com/mtuxpe/zephyr/commit/83ec3a45d79e1dbc3b64e3e0fe8311b3ece17f6b

erwango commented:
@mtuxpe 

> Yes.I need to review documentation but I changed the commit message:
Samples: Bluetooth: Add ST BLE Demo sample

Please also update the PR title and header as these are the first things that can be read on your change.

mtuxpe commented:
> @mtuxpe
> 
> > Yes.I need to review documentation but I changed the commit message:
> > Samples: Bluetooth: Add ST BLE Demo sample
> 
> Please also update the PR title and header as these are the first things that can be read on your change.

fixed

dbkinder commented on samples/bluetooth/st_ble_sensor/README.rst:
```diff
@@ -0,0 +1,59 @@
+GATT services.Currently only button notification and LED service are
+implemented.Other BLE sensor services can easily be added.
+Check `BlueST protocol`_ document.
+Although this sample can work with non stm32wb55 hardware it was tested
```
This sentence is confusing, maybe because I'm unfamiliar with the hardware names you're using.  I think you're saying you tested with a different board than listed in the example below (nucleo_wb55rg).  Can you clarify?

mtuxpe replied:
I decided to develop the demo because the original ST program for use with ST BLE Demo is quite complex.Then I found that the ST BLE demo android app  works with hardware from another manufacturer.

mtuxpe replied:
I decided to develop the demo because the original ST program for use with ST BLE Demo is quite complex.Then I found that the ST BLE demo android app works with hardware from another manufacturer.I believe ST can block access to other hardware in the future

dbkinder commented on samples/bluetooth/st_ble_sensor/README.rst:
```diff
@@ -0,0 +1,59 @@
+Instructions for `ST Nucleo WB55RG`_ board:
+-------------------------------------------
+
+This sample requests having the stm32wb5x_BLE_Stack_fw.bin binary flashed
```
I think "requests" should be "requires".  Where does someone get this ``stm32wb5x_BLE_Stack_fw.bin`` binary file?

mtuxpe replied:
fixed

mtuxpe commented on samples/bluetooth/st_ble_sensor/README.rst:
```diff
@@ -0,0 +1,59 @@
+Overview
+========
+This application demonstrates BLE peripheral by exposing vendor-specific
+GATT services.Currently only button notification and LED service are
```
fixed

mtuxpe commented on samples/bluetooth/st_ble_sensor/README.rst:
```diff
@@ -0,0 +1,59 @@
+========
+This application demonstrates BLE peripheral by exposing vendor-specific
+GATT services.Currently only button notification and LED service are
+implemented.Other BLE sensor services can easily be added.
```
fixed

mtuxpe commented on samples/bluetooth/st_ble_sensor/README.rst:
```diff
@@ -0,0 +1,59 @@
+#############################
+
+Overview
+========
```
done

dbkinder reviewed (approved):
OK, thanks.

jhedberg commented on samples/bluetooth/st_ble_sensor/src/button_svc.c:
```diff
@@ -0,0 +1,100 @@
+	LOG_INF("Button SW1 pressed");
+
+	if (p_conn) {
+
```
Unnecessary empty line here

mtuxpe replied:
> Unnecessary empty line here

fixed

jhedberg commented on samples/bluetooth/st_ble_sensor/src/button_svc.c:
```diff
@@ -0,0 +1,100 @@
+
+	if (p_conn) {
+
+		if (notify_enable == true) {
```
`notify_enable` is defined as `u8_t` so `true` is not valid. Should the variable be declared `bool` instead? (in which case explicit comparison is not necessary as it becomes a boolean expression in itself)

mtuxpe replied:
done

erwango replied:
`if (notify_enable) {`

jhedberg commented on samples/bluetooth/st_ble_sensor/src/button_svc.c:
```diff
@@ -0,0 +1,100 @@
+		if (notify_enable == true) {
+
+			err = bt_gatt_notify(NULL, &stsensor_svc->attrs[2],
+				(const void *)&but_val, sizeof(but_val));
```
Two things: the `(const void *)` case is not necessary, and the alignment of the split line seems wrong (not starting right after the ( on the first line)

mtuxpe replied:
fixed

jhedberg commented on samples/bluetooth/st_ble_sensor/src/button_svc.c:
```diff
@@ -0,0 +1,100 @@
+			err = bt_gatt_notify(NULL, &stsensor_svc->attrs[2],
+				(const void *)&but_val, sizeof(but_val));
+			if (err) {
+				LOG_ERR("Notify err");
```
How about logging the exact value of err with %d?

mtuxpe replied:
done

jhedberg commented on samples/bluetooth/st_ble_sensor/src/button_svc.c:
```diff
@@ -0,0 +1,100 @@
+				LOG_INF("Send notify ok");
+				but_val = (but_val == 0) ? 0x100 : 0;
+			}
+
```
Unnecessary empty line

mtuxpe replied:
done

jhedberg commented on samples/bluetooth/st_ble_sensor/src/button_svc.c:
```diff
@@ -0,0 +1,100 @@
+#error DT_ALIAS_SW0_GPIOS_PIN needs to be set in board.h
+#endif
+
+extern struct bt_conn *p_conn;
```
What does the p_ prefix denote here? Why not `conn`?

jhedberg commented on samples/bluetooth/st_ble_sensor/src/button_svc.c:
```diff
@@ -0,0 +1,100 @@
+		} else {
+			LOG_INF("Notify not enabled");
+		}
+
```
Unnecessary empty line

jhedberg commented on samples/bluetooth/st_ble_sensor/src/button_svc.c:
```diff
@@ -0,0 +1,100 @@
+{
+	button_dev = device_get_binding(PORT);
+	if (!button_dev)
+		return (-EOPNOTSUPP);
```
Missing { } (Zephyr requires this even for one-line branches)

jhedberg commented on samples/bluetooth/st_ble_sensor/src/led_svc.c:
```diff
@@ -0,0 +1,47 @@
+void led_on_off(u32_t led_state)
+{
+	if (led_dev)
+		gpio_pin_write(led_dev, LED, led_state);
```
Missing { }

jhedberg commented on samples/bluetooth/st_ble_sensor/src/led_svc.c:
```diff
@@ -0,0 +1,47 @@
+{
+	led_dev = device_get_binding(LED_PORT);
+	if (!led_dev)
+		return (-EOPNOTSUPP);
```
Missing { }

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,185 @@
+BT_GATT_PRIMARY_SERVICE(&st_service_uuid),
+BT_GATT_CHARACTERISTIC(&led_char_uuid.uuid,
+	BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE_WITHOUT_RESP,
+	BT_GATT_PERM_WRITE, NULL, recv, (void *)1),
```
These split lines don't seem to be correctly aligned. You should use tabs + 0-7 spaces to align starting right after the ( on the first line

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,185 @@
+	BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE_WITHOUT_RESP,
+	BT_GATT_PERM_WRITE, NULL, recv, (void *)1),
+BT_GATT_CHARACTERISTIC(&but_notif_uuid.uuid, BT_GATT_CHRC_NOTIFY,
+	BT_GATT_PERM_READ, NULL, NULL, &but_val),
```
Same here with alignment

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,185 @@
+BT_GATT_CHARACTERISTIC(&but_notif_uuid.uuid, BT_GATT_CHRC_NOTIFY,
+	BT_GATT_PERM_READ, NULL, NULL, &but_val),
+BT_GATT_CCC(mpu_ccc_cfg_changed,
+	BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
```
And here, though in this case you might be able to fit everything on a single line

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,185 @@
+		} else {
+			led_on_off(1);
+			LOG_INF("Turn on LED");
+		}
```
Add an empty line after } (whenever the next line is not also a })

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,185 @@
+			LOG_INF("Turn on LED");
+		}
+		led_state = !led_state;
+	}
```
Empty line after this

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,185 @@
+	if (err) {
+		LOG_ERR("Bluetooth init failed (err %d)", err);
+		return;
+	}
```
Empty line after this

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,185 @@
+	if (err) {
+		LOG_ERR("Advertising failed to start (err %d)", err);
+		return;
+	}
```
Empty line after this

mtuxpe replied:
ok

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,185 @@
+
+	err = button_init();
+	if (err)
+		LOG_ERR("Button init err");
```
Missing { }

Also, log the error using %d

mtuxpe replied:
ok

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,185 @@
+	/* Initialize the Bluetooth Subsystem */
+	err = bt_enable(bt_ready);
+	if (err)
+		LOG_ERR("Bluetooth init failed (err %d)", err);
```
Missing { }

mtuxpe replied:
done

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,185 @@
+	while (1) {
+		k_sleep(MSEC_PER_SEC);
+	}
+
```
Unnecessary empty line

mtuxpe replied:
done

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,185 @@
+		LOG_ERR("Bluetooth init failed (err %d)", err);
+
+	while (1) {
+		k_sleep(MSEC_PER_SEC);
```
K_SECONDS(1)

mtuxpe replied:
fixed

jhedberg reviewed (changes requested):
Mainly coding-style related comments

erwango commented on samples/bluetooth/st_ble_sensor/README.rst:
```diff
@@ -0,0 +1,59 @@
+implemented. Other BLE sensor services can easily be added.
+See `BlueST protocol`_ document for details of how to add a new service.
+
+Instructions for `ST Nucleo WB55RG`_ board
```
Please delete this section. This is not specific to this sample and valid for all BLE samples for this board.
If missing details on this board, this should be added directly in board doc.

mtuxpe replied:
ok

erwango commented on samples/bluetooth/st_ble_sensor/README.rst:
```diff
@@ -0,0 +1,59 @@
+Requirements
+************
+
+Instructions for `ST Nucleo WB55RG`_ board
```
Ditto

mtuxpe replied:
done

erwango commented on samples/bluetooth/st_ble_sensor/README.rst:
```diff
@@ -0,0 +1,59 @@
+This sample can be found under :zephyr_file:`samples/bluetooth/st_ble_sensor` in the
+Zephyr tree.
+
+Instructions for `ST Nucleo WB55RG`_ board:
```
No need to add link to extrernal doc for this board (or this should be done in board doc)

mtuxpe replied:
done

erwango commented on samples/bluetooth/st_ble_sensor/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+  description: TBD
+  name: TBD
+tests:
+  sample.bluetooth.peripheral_nucleo_wb55rg:
```
'sample.bluetooth.st_ble_sensor' ?

mtuxpe replied:
fixed

erwango commented on samples/bluetooth/st_ble_sensor/src/button_svc.c:
```diff
@@ -0,0 +1,95 @@
+
+/* GPIO button; change this to use another GPIO port */
+#ifndef DT_ALIAS_SW0_GPIOS_CONTROLLER
+#ifdef SW0_GPIO_NAME
```
Please remove the `SW0_GPIO_NAME` part.
For now, all boards should use gpio definition through dts.

mtuxpe replied:
done

erwango reviewed (changes requested):
Heads up, GPIO API being under rework, a conflict might happen with gpio code in this PR.
This will depends on PR respective landing schedules

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,187 @@
+
+static void disconnected(struct bt_conn *conn, u8_t reason)
+{
+	conn = NULL;
```
You have both a global and a local `conn`  variable, and I think the local one shadows the global here, making it pointless to try to set it to NULL. Furthermore, if this is to set the global one to NULL you need to do proper reference counting of it, i.e. `bt_conn_unref()` first.

mtuxpe replied:
fixed

jhedberg commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,187 @@
+		LOG_ERR("Connection failed (err %u)", err);
+	} else {
+		LOG_INF("Connected");
+		conn = connected;
```
You should never copy a reference counted pointer without incrementing the reference, so this should be `conn = bt_conn_ref(connected)`. Likewise, you should add a `bt_conn_unref()` when you eventually clear the global pointer (see my other comment)

mtuxpe replied:
fixed.thanks johan

erwango commented on samples/bluetooth/st_ble_sensor/README.rst:
```diff
@@ -0,0 +1,43 @@
+implemented. Other BLE sensor services can easily be added.
+See `BlueST protocol`_ document for details of how to add a new service.
+
+Requirements
```
Do we actually need this section ?

mtuxpe replied:
I think many developers don't know about the ST BLE demo app.I removed the references to stm32wb55

erwango commented on samples/bluetooth/st_ble_sensor/src/main.c:
```diff
@@ -0,0 +1,192 @@
+	LOG_INF("Notification %s", notify_enable ? "enabled" : "disabled");
+}
+
+/* The STM32WB55xx board is acting as GATT server.
```
Please remove reference to hw

mtuxpe replied:
fixed

jhedberg commented:
Will this need updating for the new GPIO API?

mtuxpe commented:
Hi Johan,
This demo uses GPIO edge interrupt and GPIO output so I think the answer is
yes.
Regards,
Marcio Montenegro

On Mon, Dec 9, 2019 at 12:12 PM Johan Hedberg <notifications@github.com>
wrote:

> Will this need updating for the new GPIO API?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/19305?email_source=notifications&email_token=ABQ7USJH6RXESDIPBBZVYMTQXZN7NA5CNFSM4IYXOZZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGJQTZA#issuecomment-563284452>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQ7USICPMG5ZIBWNA3L5GTQXZN7NANCNFSM4IYXOZZA>
> .
>

erwango commented:
@jhedberg how do we proceed vs GPIO api update?
We merge this one first and @mtuxpe push update PR on topic-gpio branch as a follow up ?

jhedberg commented:
> @jhedberg how do we proceed vs GPIO api update?
> We merge this one first and @mtuxpe push update PR on topic-gpio branch as a follow up ?

@erwango yes, that sounds like a reasonable approach to me, though then topic-gpis will need to be rebased quite soon after this one is merged.

erwango commented:
> [...] topic-gpio will need to be rebased quite soon after this one is merged.

@jhedberg, indeed, this is currently under discussion FWIU
